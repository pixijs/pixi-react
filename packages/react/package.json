{
  "name": "@pixi/react",
  "version": "7.0.2",
  "description": "Write PixiJS applications using React declarative style.",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/pixijs/pixi-react.git",
    "directory": "packages/react"
  },
  "license": "MIT",
  "sideEffects": [
    "./src/bootstrap.js"
  ],
  "main": "index.js",
  "module": "module.js",
  "jsnext:main": "module.js",
  "typings": "./dist/types/index.d.ts",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/types/index.d.ts",
        "default": "./module.js"
      },
      "require": {
        "types": "./dist/types/index.d.ts",
        "default": "./index.js"
      }
    }
  },
  "files": [
    "dist/",
    "LICENSE",
    "README.md",
    "index.d.ts",
    "index.js",
    "module.js"
  ],
  "scripts": {
    "clean": "rimraf dist index.d.ts",
    "prebuild": "npm run clean",
    "build": "npm run build:types && npm run build:dev && npm run build:prod",
    "build:types": "tsc -p tsconfig.types.json",
    "build:dev": "cross-env NODE_ENV=development rollup -c",
    "build:prod": "cross-env NODE_ENV=production rollup -c",
    "test": "jest --silent",
    "test:ci": "jest --silent --runInBand",
    "test:types": "tsc --noEmit",
    "test:watch": "jest --watch"
  },
  "devDependencies": {
    "@pixi/events": "^7.1.1",
    "@pixi/react-components": "*",
    "@pixi/react-fiber": "*",
    "@pixi/react-invariant": "*",
    "@pixi/react-modular": "*",
    "@testing-library/react": "^13.4.0",
    "jest-webgl-canvas-mock": "^0.2.3",
    "prop-types": "^15.8.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-test-renderer": "^18.2.0"
  },
  "peerDependencies": {
    "@pixi/app": ">=6.0.0",
    "@pixi/constants": ">=6.0.0",
    "@pixi/core": ">=6.0.0",
    "@pixi/display": ">=6.0.0",
    "@pixi/graphics": ">=6.0.0",
    "@pixi/math": ">=6.0.0",
    "@pixi/mesh-extras": ">=6.0.0",
    "@pixi/particle-container": ">=6.0.0",
    "@pixi/sprite": ">=6.0.0",
    "@pixi/sprite-animated": ">=6.0.0",
    "@pixi/sprite-tiling": ">=6.0.0",
    "@pixi/text": ">=6.0.0",
    "@pixi/text-bitmap": ">=6.0.0",
    "@pixi/ticker": ">=6.0.0",
    "prop-types": ">=15.0.0",
    "react": ">=17.0.0",
    "react-dom": ">=17.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "6d5b98bc6d7eafd1b0016e60a725b2b6c9f46a33"
}
