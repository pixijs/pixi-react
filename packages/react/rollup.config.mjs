import alias from '@rollup/plugin-alias';

import {
    getBuildFormat,
    getRollupConfig,
    isProductionBuild,
} from '../../shared/getRollupConfig.mjs';

const format = getBuildFormat();
const buildType = isProductionBuild() ? '' : '-dev';

const external = [
    '@babel/runtime',
    '@pixi/app',
    '@pixi/assets',
    '@pixi/constants',
    '@pixi/core',
    '@pixi/display',
    '@pixi/events',
    '@pixi/extensions',
    '@pixi/graphics',
    '@pixi/math',
    '@pixi/mesh',
    '@pixi/mesh-extras',
    '@pixi/particle-container',
    '@pixi/sprite',
    '@pixi/sprite-animated',
    '@pixi/sprite-tiling',
    '@pixi/text',
    '@pixi/text-bitmap',
    '@pixi/ticker',
    'prop-types',
    'react',
    'react-dom',
];

let builds;

if (format)
{
    builds = [
        getRollupConfig(`dist/index.${format}${buildType}.js`, format, {
            beforePlugins: [
                alias({
                    entries: {
                        '@react-spring/animated':
                            '../../shared/react-spring-create-host.js',
                    },
                }),
            ],
            external,
        }),
    ];
}
else
{
    builds = ['cjs', 'es'].map((format) =>
        getRollupConfig(`dist/index.${format}${buildType}.js`, format, {
            beforePlugins: [
                alias({
                    entries: {
                        '@react-spring/animated':
                            '../../shared/react-spring-create-host.js',
                    },
                }),
            ],
            external,
        })
    );
}

export default builds;
