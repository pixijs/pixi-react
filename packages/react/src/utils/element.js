import TYPES from '@pixi/react-tag-types';

import invariant from '../utils/invariant';
import { applyDefaultProps } from './props';
import * as components from '../components';

const ELEMENTS = Object.keys(TYPES).reduce((elements, type) => ({ ...elements, [type]: components[type] }), {});

/**
 * Inject types
 *
 * @type {Object}
 */
export const TYPES_INJECTED = {};

/**
 * Create an element based on tag type
 * Similar to react-dom's `React.createElement()`
 *
 * @param {string} type Element type
 * @param {Object} props Component props
 * @param {Object} root Root instance
 */
export function createElement(type, props = {}, root = null)
{
    const fn = ELEMENTS[type];

    let instance;
    let applyProps;

    if (typeof fn === 'function')
    {
        instance = fn(root, props);
    }

    if (!instance)
    {
    // not found, is there any injected custom component?
        const injected = TYPES_INJECTED[type];

        if (injected)
        {
            instance = injected.create(props);
            instance.didMount = injected.didMount ? injected.didMount.bind(instance) : undefined;
            instance.willUnmount = injected.willUnmount ? injected.willUnmount.bind(instance) : undefined;
            instance.applyProps = injected.applyProps ? injected.applyProps.bind(instance) : undefined;
            instance.config = injected.config;
        }
    }

    // apply initial props!
    if (instance)
    {
        applyProps = typeof instance?.applyProps === 'function' ? instance.applyProps : applyDefaultProps;
        applyProps(instance, {}, props);

        instance.__reactpixi = {
            root,
        };
    }

    return instance;
}

/**
 * Create Component
 *
 * @param {string} type
 * @param {Object} lifecycle methods
 */
export function PixiComponent(type, lifecycle)
{
    invariant(!!type, 'Expect type to be defined, got `%s`', type);
    invariant(!TYPES[type], 'Component `%s` could not be created, already exists in default components.', type);

    TYPES_INJECTED[type] = lifecycle;

    return type;
}
