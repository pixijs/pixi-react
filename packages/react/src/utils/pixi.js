import { Point, ObservablePoint } from '@pixi/math';
import invariant from '../utils/invariant';
import isNil from 'lodash.isnil';

/**
 * Parse PIXI point value to array of coordinates
 *
 * @param {*} value
 * @returns {Array}
 */
export function parsePoint(value)
{
    let arr = [];

    if (typeof value === 'undefined')
    {
        return arr;
    }
    else if (typeof value === 'string')
    {
        arr = value.split(',');
    }
    else if (typeof value === 'number')
    {
        arr = [value];
    }
    else if (Array.isArray(value))
    {
        arr = [...value];
    }
    else if (value !== null && typeof value === 'object')
    {
        const x = (value && value?.x) || 0;
        const y = (value && value?.y) || 0;

        arr = [x, y];
    }
    else
    {
        return arr;
    }

    return arr.filter((p) => !isNil(p) && !isNaN(p)).map(Number);
}

/**
 * Check if two points are equal
 *
 * @param {*} oldValue
 * @param {*} newValue
 * @returns {boolean}
 */
export function pointsAreEqual(oldValue, newValue)
{
    if (oldValue.length !== newValue.length)
    {
        return false;
    }

    for (let i = 0; i < oldValue.length; i++)
    {
        if (oldValue[i] !== newValue[i])
        {
            return false;
        }
    }

    return true;
}

/**
 * Determine value is type of Point or ObservablePoint
 * See https://github.com/michalochman/react-pixi-fiber/blob/a4dbddbef0ffbf0f563c64d30766ea28222a51ea/src/utils.js#L48
 *
 * @param {*} value
 * @returns {boolean}
 */
export function isPointType(value)
{
    return value instanceof Point || value instanceof ObservablePoint;
}

/**
 * Event handlers
 *
 * @type {string[]}
 */
export const eventHandlers = [
    'click',
    'mousedown',
    'mousemove',
    'mouseout',
    'mouseover',
    'mouseup',
    'mouseupoutside',
    'tap',
    'touchstart',
    'touchmove',
    'touchend',
    'touchendoutside',
    'pointercancel',
    'pointerout',
    'pointerover',
    'pointertap',
    'pointerdown',
    'pointerup',
    'pointerupoutside',
    'pointermove',
    'rightclick',
    'rightdown',
    'rightup',
    'rightupoutside',
    'touchcancel',
    'touchendoutside',
    'touchmove',
    'touchstart',
];

/**
 * Set value on a PIXI.DisplayObject
 * See https://github.com/Izzimach/react-pixi/blob/a25196251a13ed9bb116a8576d93e9fceac2a14c/src/ReactPIXI.js#L114
 *
 * @param {PIXI.DisplayObject} instance
 * @param {string} prop
 * @param {*} value
 */
export function setValue(instance, prop, value)
{
    if (isPointType(instance[prop]) && isPointType(value))
    {
    // copy value
        instance[prop].copyFrom(value);
    }
    else if (isPointType(instance[prop]))
    {
    // parse value if a non-Point type is being assigned to a Point type
        const coordinates = parsePoint(value);

        invariant(
            typeof coordinates !== 'undefined' && coordinates.length > 0 && coordinates.length < 3,
            'The property `%s` is a `PIXI.Point` or `PIXI.ObservablePoint` and must be set to a comma-separated string of '
        + 'either 1 or 2 coordinates, a 1 or 2 element array containing coordinates, or a PIXI Point/ObservablePoint. '
        + 'If only one coordinate is given then X and Y will be set to the provided value. Received: `%s` of type `%s`.',
            prop,
            JSON.stringify(value),
            typeof value
        );

        instance[prop].set(coordinates.shift(), coordinates.shift());
    }
    else
    {
    // just hard assign value
        instance[prop] = value;
    }
}
