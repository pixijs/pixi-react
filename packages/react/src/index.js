import { TYPES } from './utils/element';

/**
 * -------------------------------------------
 * Public API
 * -------------------------------------------
 */

export * from './exports';

export const BitmapText = TYPES.BitmapText;
export const Container = TYPES.Container;
export const Graphics = TYPES.Graphics;
export const NineSlicePlane = TYPES.NineSlicePlane;
export const ParticleContainer = TYPES.ParticleContainer;
export const Sprite = TYPES.Sprite;
export const AnimatedSprite = TYPES.AnimatedSprite;
export const Text = TYPES.Text;
export const TilingSprite = TYPES.TilingSprite;
export const SimpleMesh = TYPES.SimpleMesh;
export const SimpleRope = TYPES.SimpleRope;
