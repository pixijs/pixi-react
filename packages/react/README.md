<p align="center">
  <img src="https://user-images.githubusercontent.com/232559/142733492-3c106f68-8b8e-459c-95f9-aca77561d438.png" alt="pixi-react" width="310" />
</p>

<h1 align="center">
  Pixi React
</h1>

<p align="center">
  <strong>Simply the best way to write PIXI applications in React</strong>
  <br />
  <sub>Write <a href="http://www.pixijs.com/">PIXI</a> applications using React declarative style ðŸ‘Œ</sub>
</p>

<br />

<p align="center">
  <img src="https://img.shields.io/github/v/release/pixijs/pixi-react" alt="release" />
  <img src="https://img.shields.io/npm/dm/@pixi/react" alt="downloads" />
  <img src="https://img.shields.io/circleci/project/github/pixijs/pixi-react/master.svg" alt="ci tests" />
  <img src="https://img.shields.io/badge/license-MIT-green.svg" alt="license" />
  <img src="https://img.shields.io/badge/react-latest-ff69b4.svg" alt="react version" />
  <img src="https://img.shields.io/badge/pixi-v6+-ff69b4.svg" alt="pixi version" />
</p>

<br />

Pixi React is an open-source, production-ready library to render high performant PIXI applications in React.

## News!

We're delighted to announce Pixi React v7.0.0, the first major release since making it an official Pixi package!

While all the core components remain the same, there have been some significant changes under the hood:

* Full React 18 support including a new `createRoot` API, matching the signature of [React 18's `ReactDOM/client`](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis)
* Full Pixi V7 support
* New up-to-date [docs site](https://pixijs.io/pixi-react/)
* Repository converted to a monorepo with separate [`@pixi/react`](https://www.npmjs.com/package/@pixi/react) and [`@pixi/animated`](https://www.npmjs.com/package/@pixi/react-animated) packages
* Internally all `pixi.js` library imports were replaced with `@pixi/` scoped package versions, providing bundlesize improvements for users who use modular builds
* Scoped imports allowed us to remove `@pixi/react-legacy` and `@pixi/react-animated-legacy` packages, with rendering deferred to a user's installed `pixi.js` package
* `sideEffects: false` added to `package.json` to support tree-shaking

We're excited to see what the community builds with the library and as ever please let us know on [GitHub](https://github.com/pixijs/pixi-react/issues) if you run into any issues, or reach out to us on the [Discord](https://discord.com/channels/734147990985375826/968068526566965279) to chat.

Thanks!
PixiJS Team

## Get started

### Quick start

If you want to start a new React project from scratch, we recommend [Create React App](https://github.com/facebook/create-react-app).
To add to an existing React application, just install the dependencies:

#### Start New React Project "my-app" with Create React App:
```bash
# for typescript add "--template typescript"
npx create-react-app my-app
cd my-app
```

#### Install Pixi React Dependencies:
```bash
npm install pixi.js @pixi/react
```

#### Usage:
```jsx
import { BlurFilter } from 'pixi.js';
import { Stage, Container, Sprite, Text } from '@pixi/react';
import { useMemo } from 'react';

export const MyComponent = () =>
{
  const blurFilter = useMemo(() => new BlurFilter(4), []);

  return (
    <Stage>
      <Sprite
        image="https://pixijs.io/pixi-react/img/bunny.png"
        x={400}
        y={270}
        anchor={{ x: 0.5, y: 0.5 }}
      />

      <Container x={400} y={330}>
        <Text text="Hello World" anchor={{ x: 0.5, y: 0.5 }} filters={[blurFilter]} />
      </Container>
    </Stage>
  );
};
```

### Docs

Check out [our documentation](https://pixijs.io/pixi-react/) for guides and a full API reference.

Or checkout [our examples](https://codepen.io/collection/XPpGdb) on codepen for inspiration.

## Contribute

Want to contribute to Pixi React? Our [contributing guide](/CONTRIBUTING.md) has you covered.

## License

Pixi React is MIT licensed.

This project was originally created and maintained by @inlet. He graciously gave the Pixi team the keys to maintain and push the project forward. Big thanks for all the hard work on making this project so awesome.

This projects codebase was originally forked from @michalochman 's [react-pixi-fiber](https://github.com/michalochman/react-pixi-fiber).

## Meet fellow developers

You have an amazing feature in mind or just want to get in touch with other developers? Feel free to join our Discord channel.

[Join us on Discord](https://discord.gg/zqbXQAADuM)
