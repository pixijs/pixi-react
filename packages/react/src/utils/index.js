import * as element from './element';
import * as props from './props';
import * as pixi from './pixi';

export { element, props, pixi };
