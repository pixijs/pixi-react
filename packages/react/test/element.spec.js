import * as PIXI from 'pixi.js';
import React from 'react';
import TYPES from '@pixi/react-tag-types';
import { render } from '@testing-library/react';

import { createElement, TYPES_INJECTED, PixiComponent } from '../src/utils/element';
import Stage from '../src/stage';

import { emptyTexture } from './__fixtures__/textures';
import { desyrel } from './__fixtures__/bitmapfonts';
import parseBitmapFont from './__utils__/parseBitmapFont';

parseBitmapFont(desyrel);

describe('createElement', () =>
{
    test('types', () =>
    {
        expect(TYPES).toMatchSnapshot();
    });

    test('create Container', () =>
    {
        const element = createElement(TYPES.Container);

        expect(element).toBeInstanceOf(PIXI.Container);
    });

    test('create Text', () =>
    {
        const element = createElement(TYPES.Text, { text: 'foobar' });

        expect(element).toBeInstanceOf(PIXI.Text);
    });

    test('create Text as a sprite', () =>
    {
        const element = createElement(TYPES.Text, { text: 'foobar', isSprite: true });

        expect(element).toBeInstanceOf(PIXI.Sprite);
    });

    test('create Sprite', () =>
    {
        const element = createElement(TYPES.Sprite, { texture: emptyTexture });

        expect(element).toBeInstanceOf(PIXI.Sprite);
    });

    test('create AnimatedSprite', () =>
    {
        const element = createElement(TYPES.AnimatedSprite, { textures: [emptyTexture] });

        expect(element).toBeInstanceOf(PIXI.AnimatedSprite);
    });

    test('create ParticleContainer', () =>
    {
        const element = createElement(TYPES.ParticleContainer);

        expect(element).toBeInstanceOf(PIXI.ParticleContainer);
    });

    test('create BitmapText', () =>
    {
        const element = createElement(TYPES.BitmapText, {
            text: 'foobar',
            style: { fontName: 'Desyrel', fontSize: 35, align: 'left' },
        });

        expect(element).toBeInstanceOf(PIXI.BitmapText);
    });

    test('create TilingSprite', () =>
    {
        const element = createElement(TYPES.TilingSprite, { texture: emptyTexture });

        expect(element).toBeInstanceOf(PIXI.TilingSprite);
    });

    test('create Graphics', () =>
    {
        const element = createElement(TYPES.Graphics);

        expect(element).toBeInstanceOf(PIXI.Graphics);
    });

    test('create NineSlicePlane', () =>
    {
        const element = createElement(TYPES.NineSlicePlane, { texture: emptyTexture });

        expect(element).toBeInstanceOf(PIXI.NineSlicePlane);
    });

    test('create SimpleMesh', () =>
    {
        const element = createElement(TYPES.SimpleMesh, { texture: emptyTexture });

        expect(element).toBeInstanceOf(PIXI.SimpleMesh);
    });

    test('create SimpleRope', () =>
    {
        const element = createElement(TYPES.SimpleRope, {
            texture: emptyTexture,
            points: [new PIXI.Point(0, 0), new PIXI.Point(20, 20)],
        });

        expect(element).toBeInstanceOf(PIXI.SimpleRope);
    });

    test('get undefined', () =>
    {
        expect(createElement('INVALID')).toBeUndefined();
    });
});

describe('element.applyProps', () =>
{
    let spy;

    beforeAll(() =>
    {
        spy = jest.spyOn(PIXI.Texture, 'from').mockReturnValue(emptyTexture);
    });

    afterAll(() =>
    {
        spy.mockRestore();
    });

    test('Sprite.applyProps exists', () =>
    {
        const element = createElement(TYPES.Sprite, { image: './image.png' });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toHaveBeenCalledWith('./image.png');
    });

    test('AnimatedSprite.applyProps with images prop exists', () =>
    {
        const element = createElement(TYPES.AnimatedSprite, { images: ['./image.png'] });

        expect(element).toHaveProperty('applyProps');
        expect(spy).lastCalledWith('./image.png');
    });

    test('AnimatedSprite.applyProps with textures prop exists', () =>
    {
        const element = createElement(TYPES.AnimatedSprite, { textures: [PIXI.Texture.from('./image.png')] });

        expect(element).toHaveProperty('applyProps');
        expect(spy).lastCalledWith('./image.png');
    });

    test('Sprite.applyProps image', () =>
    {
        const element = createElement(TYPES.Sprite, { image: './image.png' });

        expect(spy).lastCalledWith('./image.png');

        const changed = element.applyProps(element, { image: './image.png' }, { image: './new-image.png' });

        expect(spy).lastCalledWith('./new-image.png');
        expect(changed).toBeFalsy();
    });

    test('Sprite.applyProps texture', () =>
    {
        const element = createElement(TYPES.Sprite, { texture: emptyTexture });

        const changed = element.applyProps(element, { texture: emptyTexture }, { image: './image.png' });

        expect(changed).toBeTruthy();
    });

    test('TilingSprite.applyProps exists', () =>
    {
        const element = createElement(TYPES.TilingSprite, { image: './image.png' });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toHaveBeenCalledWith('./image.png');
    });

    test('TilingSprite.applyProps image', () =>
    {
        const element = createElement(TYPES.TilingSprite, { image: './image.png' });

        expect(spy).lastCalledWith('./image.png');

        const changed = element.applyProps(element, { image: './image.png' }, { image: './new-image.png' });

        expect(changed).toBeFalsy();
        expect(spy).lastCalledWith('./new-image.png');
    });

    test('TilingSprite.applyProps texture', () =>
    {
        const element = createElement(TYPES.TilingSprite, { texture: emptyTexture });

        const changed = element.applyProps(element, { texture: emptyTexture }, { image: './image.png' });

        expect(changed).toBeTruthy();
    });

    test('TilingSprite.applyProps tilePosition', () =>
    {
        const oldPosition = '1, 2';
        const newPosition = { x: 12, y: 20 };
        const element = createElement(TYPES.TilingSprite, { tilePosition: oldPosition, image: './image.png' });

        const changed = element.applyProps(
            element,
            { tilePosition: oldPosition, image: './image.png' },
            { tilePosition: newPosition, image: './image.png' }
        );

        expect(changed).toBeTruthy();
    });

    test('SimpleRope.applyProps exists', () =>
    {
        const element = createElement(TYPES.SimpleRope, {
            image: './image.png',
            points: [new PIXI.Point(0, 0), new PIXI.Point(20, 20)],
        });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toHaveBeenCalledWith('./image.png');
    });

    test('SimpleRope.applyProps image', () =>
    {
        const element = createElement(TYPES.SimpleRope, {
            image: './image.png',
            points: [new PIXI.Point(0, 0), new PIXI.Point(20, 20)],
        });

        expect(spy).lastCalledWith('./image.png');

        const changed = element.applyProps(
            element,
            { image: './image.png' },
            {
                image: './new-image.png',
                points: [new PIXI.Point(0, 0), new PIXI.Point(20, 20)],
            }
        );

        expect(spy).lastCalledWith('./new-image.png');
        expect(changed).toBeTruthy();
    });

    test('NineSlicePlane.applyProps exists', () =>
    {
        const element = createElement(TYPES.NineSlicePlane, { image: './image.png' });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toHaveBeenCalledWith('./image.png');
    });

    test('NineSlicePlane.applyProps image', () =>
    {
        const element = createElement(TYPES.NineSlicePlane, { image: './image.png' });

        expect(spy).lastCalledWith('./image.png');

        const changed = element.applyProps(element, { image: './image.png' }, { image: './new-image.png' });

        expect(spy).lastCalledWith('./new-image.png');
        expect(changed).toBeFalsy();
    });

    test('NineSlicePlane.applyProps texture', () =>
    {
        const element = createElement(TYPES.NineSlicePlane, { texture: emptyTexture });
        const changed = element.applyProps(element, { texture: emptyTexture }, { image: './new-image.png' });

        expect(changed).toBeTruthy();
    });

    test('SimpleMesh.applyProps exists', () =>
    {
        const element = createElement(TYPES.SimpleMesh, { image: './image.png' });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toHaveBeenCalledWith('./image.png');
    });

    test('SimpleMesh.applyProps image', () =>
    {
        const element = createElement(TYPES.SimpleMesh, { image: './image.png' });

        expect(spy).lastCalledWith('./image.png');

        const changed = element.applyProps(element, { image: './image.png' }, { image: './new-image.png' });

        expect(spy).lastCalledWith('./new-image.png');
        expect(changed).toBeFalsy();
    });

    test('SimpleMesh.applyProps texture', () =>
    {
        const element = createElement(TYPES.SimpleMesh, { texture: emptyTexture });

        const changed = element.applyProps(element, { texture: emptyTexture }, { image: './new-image.png' });

        expect(changed).toBeTruthy();
    });

    test('Graphics.applyProps exists', () =>
    {
        const spy = jest.fn();

        const element = createElement(TYPES.Graphics, { draw: spy });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toBeCalledWith(element);
    });

    test('Graphics.applyProps draw', () =>
    {
        const spy = jest.fn();

        const element = createElement(TYPES.Graphics, { draw: spy });

        expect(spy).toHaveBeenCalledTimes(1);

        const applied = element.applyProps(element, { draw: spy }, { draw: spy });

        expect(spy).toHaveBeenCalledTimes(1);
        expect(applied).toBeFalsy();
    });

    test('Graphics.applyProps draw prevented twice', () =>
    {
        const draw1 = jest.fn();
        const draw2 = jest.fn();
        const props = { draw: draw1 };
        const nextProps = { draw: draw2 };
        let applied = false;
        const element = createElement(TYPES.Graphics, props);

        applied = element.applyProps(element, props, props);
        expect(applied).toBeFalsy();
        applied = element.applyProps(element, props, props);
        expect(applied).toBeFalsy();
        expect(draw1).toHaveBeenCalledTimes(1);
        applied = element.applyProps(element, props, nextProps);
        expect(applied).toBeTruthy();
        applied = element.applyProps(element, nextProps, nextProps);
        expect(applied).toBeFalsy();
        expect(draw2).toHaveBeenCalledTimes(1);
    });
});

describe('PixiComponent', () =>
{
    afterEach(() =>
    {
        Object.keys(TYPES_INJECTED).forEach((k) =>
        {
            delete TYPES_INJECTED[k];
        });
    });

    test('type must be defined', () =>
    {
        expect(() => new PixiComponent(null)).toThrow('Expect type to be defined, got `null`');
    });

    test('cannot override existing component', () =>
    {
        expect(() => new PixiComponent('Text', {})).toThrow(
            'Component `Text` could not be created, already exists in default components.'
        );
    });

    test('inject custom component', () =>
    {
        const lifecycle = { create: (props) => {} };

        // eslint-disable-next-line no-new
        new PixiComponent('Rectangle', lifecycle);
        expect(TYPES_INJECTED).toHaveProperty('Rectangle', lifecycle);
    });

    test('create injected component', () =>
    {
        const scoped = jest.fn();
        const config = {
            destroyChildren: true,
            destroy: true
        };

        const lifecycle = {
            create: jest.fn(() => new PIXI.Graphics()),
            didMount: jest.fn(),
            willUnmount: jest.fn(),
            applyProps: jest.fn(function applyProps()
            {
                scoped(this);
            }),
            config
        };

        // eslint-disable-next-line no-new
        new PixiComponent('Rectangle', lifecycle);

        const props = { x: 100, y: 200 };
        const element = createElement('Rectangle', props);

        expect(element.didMount).toBeDefined();
        expect(element.willUnmount).toBeDefined();
        expect(element.applyProps).toBeDefined();
        expect(element.config).toBe(config);
        expect(element).toBeInstanceOf(PIXI.Graphics);
        expect(lifecycle.create).toHaveBeenCalledTimes(1);
        expect(lifecycle.create).toHaveBeenCalledWith(props);
        expect(lifecycle.applyProps).toHaveBeenCalledTimes(1);
        expect(scoped).toHaveBeenCalledTimes(1);
        expect(scoped).toHaveBeenCalledWith(element);
    });

    test.each`
        destroyChildren
        ${true}
        ${false}
    `('unmount and destroy component when destroyChildren is $destroyChildren', ({ destroyChildren }) =>
    {
        const scoped = jest.fn();
        const config = {
            destroyChildren,
            destroy: true
        };

        const makeLifecycle = (pixiInstance) => ({
            create: jest.fn(() => pixiInstance),
            didMount: jest.fn(),
            willUnmount: jest.fn(),
            applyProps: jest.fn(function applyProps()
            {
                scoped(this);
            }),
            config
        });

        const parentPixiInstance = new PIXI.Container();
        const childPixiInstance = new PIXI.Container();

        const parentDestroySpy = jest.spyOn(parentPixiInstance, 'destroy');
        const childDestroySpy = jest.spyOn(childPixiInstance, 'destroy');

        const parentLifecycle = makeLifecycle(parentPixiInstance);
        const childLifecycle = makeLifecycle(childPixiInstance);

        const Parent = PixiComponent('Parent', parentLifecycle);
        const Child = PixiComponent('Child', childLifecycle);

        const { unmount } = render(
            <Stage width={800} height={600}>
                <Parent>
                    <Child />
                </Parent>
            </Stage>
        );

        expect(parentLifecycle.create).toHaveBeenCalledTimes(1);
        expect(parentLifecycle.didMount).toHaveBeenCalledTimes(1);
        expect(parentLifecycle.applyProps).toHaveBeenCalledTimes(1);
        expect(childLifecycle.create).toHaveBeenCalledTimes(1);
        expect(childLifecycle.didMount).toHaveBeenCalledTimes(1);
        expect(childLifecycle.applyProps).toHaveBeenCalledTimes(1);

        unmount();

        expect(parentLifecycle.willUnmount).toHaveBeenCalledTimes(1);
        expect(parentDestroySpy).toHaveBeenCalledTimes(1);
        expect(parentDestroySpy).toHaveBeenCalledWith({
            children: destroyChildren,
            texture: false,
            baseTexture: false
        });

        if (destroyChildren)
        {
            expect(childLifecycle.willUnmount).toHaveBeenCalledTimes(1);
            expect(childDestroySpy).toHaveBeenCalledTimes(1);
            expect(childDestroySpy).toHaveBeenCalledWith({
                children: destroyChildren,
                texture: false,
                baseTexture: false
            });
        }
        else
        {
            expect(childLifecycle.willUnmount).not.toHaveBeenCalled();
            expect(childDestroySpy).not.toHaveBeenCalled();
        }
    });

    test('create injected component without lifecycle methods', () =>
    {
        // eslint-disable-next-line no-new
        new PixiComponent('Rectangle', {
            create: () => new PIXI.Graphics(),
        });

        const element = createElement('Rectangle');

        expect(element.didMount).toBeUndefined();
        expect(element.willUnmount).toBeUndefined();
        expect(element.applyProps).toBeUndefined();
    });
});
