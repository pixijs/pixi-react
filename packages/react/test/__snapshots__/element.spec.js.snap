// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createElement types 1`] = `
{
  "AnimatedSprite": "AnimatedSprite",
  "BitmapText": "BitmapText",
  "Container": "Container",
  "Graphics": "Graphics",
  "NineSlicePlane": "NineSlicePlane",
  "ParticleContainer": "ParticleContainer",
  "SimpleMesh": "SimpleMesh",
  "SimpleRope": "SimpleRope",
  "Sprite": "Sprite",
  "Text": "Text",
  "TilingSprite": "TilingSprite",
}
`;
